plugin/pathvirt/pathvirt.cpp

```diff
@@ -0,0 +1,224 @@
+   is set on restart */
+static int shouldSwap;
+
+/* NOTE: DMTCP_PATH_PREFIX env variables cannot exceed MAX_ENV_VAR_SIZE
+   characters in length */
+static char oldPathPrefixList[MAX_ENV_VAR_SIZE];
+static char newPathPrefixList[MAX_ENV_VAR_SIZE];
+
+/*
+ * Helper Functions
+ */
+
+/*
+ * clfind - returns first index in colonList which is a prefix for path
+ *          modifies the @listPtr to point to the element in colonList
+ */
+static int
+clfind(const char *colonList,  // IN
+       const char *path,       // IN
+       char **listPtr)         // OUT
+{
+    int index = 0;
+    char *element = const_cast<char *>(colonList);
+    char *colon = NULL;
+
+    /* while there is a colon present, loop */
+    while (colon = strchr(element, ':')) {
+        /* check if element is a prefix of path. here, colon - element is
+           an easy way to calculate the length of the element in the list
+           to use as the size parameter to strncmp */
+        if (strncmp(path, element, colon - element) == 0) {
+            *listPtr = element;
+            return index;
+        }
+
+        /* move element to point to next element */
+        element = colon + 1;
+
+        index++;
+    }
+
+    /* process the last element in the list */
+    if (strncmp(path, element, strlen(element)) == 0) {
+        *listPtr = element;
+        return index;
+    }
+
+    /* not found */
+    return -1;
+}
+
+/*
+ * clget - returns pointer to element in colonList at index i
+ *         and NULL if not found
+ */
+static char*
+clget(const char *colonList, unsigned int i)
+{
+    int curr_ind = 0;
+    char *element = const_cast<char *>(colonList);
+    char *colon = NULL;
+
+    /* iterate through elements until last one */
+    while (colon = strchr(element, ':')) {
+        /* if we are at the request index, return pointer to start of element */
+        if (curr_ind == i)
+            return element;
+
+        /* otherwise, advance pointer to next element and bump current index */
+        element = colon + 1;
+        curr_ind++;
+    }
+
+    /* last element */
+    if (curr_ind == i)
+        return element;
+
+    /* not found */
+    return NULL;
+}
+
+/*
+ * clgetsize - returns size of an element pointed to by @element in the
+ *             list
+ */
+static size_t
+clgetsize(const char *colonList, const char *element)
+{
+    /* either calculate the element's length, or call
+     * strlen if element was last one */
+    const char *colon = strchr(element, ':');
+    return colon ? colon - element : strlen(element);
+}
+
+/*
+ * virtual_to_physical_path - translate virtual to physical path
+ *
+ * Returns a const char* for the corresponding physical path to the given
+ * virtual path. If no path translation occurred, the given virtual path
+ * will simply be returned. If path translation *did* occur, the translated
+ * physical path will also be available as a dmtcp::string from the second
+ * parameter.
```

Comment:
Just a short suggested additional comment:
```
 *     Conceptually, an original path prior to the first checkpoint is considered a "virtual path".
 * After a restart, it will be substituted using the latest list of registered paths.  Hence,
 * a newly registered path to be substituted is a "physical path".  Internally, DMTCP works
 * with the original "virtual path" as the canonical name.  But in any system calls, it must
 * translate the virtual path to the latest "physical path", which will correspond to the
 * current post-restart filesystem.
 */
```